File: backend/application/dto/vehicle_position.py
from pydantic import BaseModel, Field

from backend.domain.gtfs_rt.enums import City
from backend.domain.gtfs_rt.vehicle_position import VehiclePosition


class VehiclePathDTO(BaseModel):
    city: City = Field(description="City")
    route_id: str = Field(description="Route ID")


class VehiclePositionDTO(BaseModel):
    id: str = Field(description="Vehicle ID")
    trip_id: str = Field(description="Trip ID")
    route_id: str = Field(description="Route ID")
    direction_id: int = Field(description="Direction ID")
    latitude: float = Field(description="Latitude")
    longitude: float = Field(description="Longitude")
    bearing: int = Field(description="Heading in degrees")
    speed: int = Field(description="Speed")
    current_status: str = Field(description="Vehicle stop status")
    timestamp: int = Field(description="Position timestamp")

    @classmethod
    def from_domain(cls, position: VehiclePosition) -> "VehiclePositionDTO":
        return cls(
            id=position.id,
            trip_id=position.trip.id,
            route_id=position.trip.route_id,
            direction_id=position.trip.direction_id,
            latitude=position.position.latitude,
            longitude=position.position.longitude,
            bearing=position.position.bearing,
            speed=position.position.speed,
            current_status=position.current_status,
            timestamp=position.timestamp,
        )

File: frontend/lib/infrastructure/backend/models/response/vehicle_position.dart
import 'package:frontend/domain/vehicle_position.dart' show VehiclePosition;


class VehiclePositionResponse {
    final String id;
    final String tripId;
    final String routeId;
    final int directionId;
    final double latitude;
    final double longitude;
    final int bearing;
    final int speed;
    final String currentStatus;
    final int timestamp;

    VehiclePositionResponse({
        required this.id,
        required this.tripId,
        required this.routeId,
        required this.directionId,
        required this.latitude,
        required this.longitude,
        required this.bearing,
        required this.speed,
        required this.currentStatus,
        required this.timestamp,
    });

    factory VehiclePositionResponse.fromJson(Map<String, dynamic> json) =>
        VehiclePositionResponse(
            id: json['id'] as String,
            tripId: json['trip_id'] as String,
            routeId: json['route_id'] as String,
            directionId: (json['direction_id'] as num).toInt(),
            latitude: (json['latitude'] as num).toDouble(),
            longitude: (json['longitude'] as num).toDouble(),
            bearing: (json['bearing'] as num).toInt(),
            speed: (json['speed'] as num).toInt(),
            currentStatus: json['current_status'] as String? ?? '',
            timestamp: (json['timestamp'] as num).toInt(),
        );

    VehiclePosition toDomain() => VehiclePosition(
        id: id,
        tripId: tripId,
        routeId: routeId,
        directionId: directionId,
        latitude: latitude,
        longitude: longitude,
        bearing: bearing,
        speed: speed,
        currentStatus: currentStatus,
        timestamp: timestamp,
    );
}
